src/main/java/relations/r1cs/FileToR1CS.java

```diff
@@ -0,0 +1,478 @@
+            auxInputs = (JSONArray) jsonObject.get("aux_input");
+            constraintList = (JSONArray) jsonObject.get("constraints");
+
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        // TODO - do we really need to load this into an ArrayList?
+        int numConstraints = constraintList.size();
+        JSONArray[] constraintArray = new JSONArray[numConstraints];
+        for (int i = 0; i < numConstraints; i++) {
+            constraintArray[i] = (JSONArray) constraintList.get(i);
+        }
+
+        // instantiate partition array
+        final int numPartitions = config.numPartitions();
+        final ArrayList<Integer> partitions = new ArrayList<>();
+        for (int i = 0; i < numPartitions; i++) {
+            partitions.add(i);
+        }
+        if (numConstraints % 2 != 0) {
+            partitions.add(numPartitions);
+        }
+
+        // Load Linear Combinations as RDD format
+        JavaPairRDD<Long, LinearTerm<FieldT>> linearCombinationA = makeDistributedCombinationFromJSON(
+                config, partitions, constraintArray, 0, numPartitions, numConstraints);
+
+        JavaPairRDD<Long, LinearTerm<FieldT>> linearCombinationB = makeDistributedCombinationFromJSON(
+                config, partitions, constraintArray, 1, numPartitions, numConstraints);
+
+        JavaPairRDD<Long, LinearTerm<FieldT>> linearCombinationC = makeDistributedCombinationFromJSON(
+                config, partitions, constraintArray, 2, numPartitions, numConstraints);
+
+        // Serial assignment may not be necessary.
+        final Assignment<FieldT> serialAssignment = new Assignment<>();
+        int numInputs = primaryInputs.size();
+        for (int i = 0; i < numInputs; i++) {
+            final BN254aFields.BN254aFr value = new BN254aFields.BN254aFr((String) primaryInputs.get(i));
+            serialAssignment.add((FieldT) value);
+        }
+        int numAuxiliary = auxInputs.size();
+        for (int i = 0; i < numAuxiliary; i++) {
+            final BN254aFields.BN254aFr value = new BN254aFields.BN254aFr((String) auxInputs.get(i));
+            serialAssignment.add((FieldT) value);
+        }
+
+        final long numVariables = numInputs + numAuxiliary;
+
+        final ArrayList<Integer> variablePartitions = new ArrayList<>();
+        for (int i = 0; i < numPartitions; i++) {
+            variablePartitions.add(i);
+        }
+        if (numVariables % 2 != 0) {
+            variablePartitions.add(numPartitions);
+        }
+
+        final int numExecutors = config.numExecutors();
+        final ArrayList<Integer> assignmentPartitions = new ArrayList<>();
+        for (int i = 0; i < numExecutors; i++) {
+            assignmentPartitions.add(i);
+        }
+        if (numVariables % 2 != 0) {
+            assignmentPartitions.add(numExecutors);
+        }
+
+        JavaPairRDD<Long, FieldT> oneFullAssignment = config.sparkContext()
```

Comment:
I'm not sure if that's possible, but I'd just pass in the primary and auxiliary inputs here and build the assignment  dynamically inside the `parallelize` function instead of creating it twice.

Reply:
Yes I totally agree. This was only my first pass through the implementation and I intend to clean this up ASAP.

Reply:
Is there a way to concatenate two lists without iterating over them?